File: ocr_reader.py
# ocr_reader.py
import sys
import os
import argparse
from PIL import Image
import pytesseract
from docx import Document

def extract_text(image_path, lang='eng'):
    """Extract text from image using Tesseract OCR."""
    try:
        img = Image.open(image_path)
        text = pytesseract.image_to_string(img, lang=lang)
        return text.strip()
    except Exception as e:
        print(f"Error processing {image_path}: {e}", file=sys.stderr)
        return None

def export_docx(text, output_path):
    """Export extracted text to DOCX format."""
    doc = Document()
    doc.add_heading('OCR Extracted Text', 0)
    doc.add_paragraph(text)
    doc.save(output_path)
    return output_path

def export_txt(text, output_path):
    """Export extracted text to plain text format."""
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(text)
    return output_path

def main():
    parser = argparse.ArgumentParser(description="OCR Reader")
    parser.add_argument('images', nargs='+', help='Image files to process')
    parser.add_argument('--output', default='output.docx', help='Output file')
    parser.add_argument('--lang', default='eng', help='OCR language (default: eng)')
    parser.add_argument('--format', choices=['docx', 'txt'], default='docx', help='Output format')
    args = parser.parse_args()

    print("📄 OCR Reader")
    print(f"Images: {', '.join(args.images)}")
    print(f"Language: {args.lang}")

    all_text = []
    for img_path in args.images:
        print(f"Processing: {img_path}")
        text = extract_text(img_path, args.lang)
        if text:
            all_text.append(f"--- {img_path} ---\n{text}\n")
        else:
            print(f"⚠️ No text extracted from {img_path}")

    if not all_text:
        print("❌ No text extracted from any image.")
        sys.exit(1)

    combined_text = "\n".join(all_text)
    print("\n📖 Extracted text:")
    print(combined_text[:500] + ("..." if len(combined_text) > 500 else ""))

    if args.format == 'docx':
        output_path = export_docx(combined_text, args.output)
        print(f"\n📄 Exporting to DOCX... ✅")
    else:
        output_path = export_txt(combined_text, args.output)
        print(f"\n📄 Exporting to TXT... ✅")
    print(f"File saved: {output_path}")

if __name__ == "__main__":
    main()

File: ocr_reader.go
// ocr_reader.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"

	"github.com/nguyenthenguyen/docx"
	"github.com/otiai10/gosseract/v2"
)

func extractText(imagePath string, lang string) (string, error) {
	client := gosseract.NewClient()
	defer client.Close()
	client.SetLanguage(lang)
	client.SetImage(imagePath)
	text, err := client.Text()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(text), nil
}

func exportDocx(text, outputPath string) error {
	doc := docx.NewDocument()
	doc.AddHeading("OCR Extracted Text", 0)
	doc.AddParagraph(text)
	return doc.Save(outputPath)
}

func exportTxt(text, outputPath string) error {
	return os.WriteFile(outputPath, []byte(text), 0644)
}

func main() {
	var (
		images  = flag.String("images", "", "Comma-separated image files")
		output  = flag.String("output", "output.docx", "Output file")
		lang    = flag.String("lang", "eng", "OCR language")
		format  = flag.String("format", "docx", "Output format (docx|txt)")
	)
	flag.Parse()

	if *images == "" && len(flag.Args()) > 0 {
		*images = flag.Args()[0]
	}
	if *images == "" {
		fmt.Println("Usage: ocr_reader -images image1.png,image2.jpg [options]")
		os.Exit(1)
	}

	imageList := strings.Split(*images, ",")
	fmt.Println("📄 OCR Reader")
	fmt.Printf("Images: %s\n", *images)
	fmt.Printf("Language: %s\n", *lang)

	var allText []string
	for _, imgPath := range imageList {
		fmt.Printf("Processing: %s\n", imgPath)
		text, err := extractText(strings.TrimSpace(imgPath), *lang)
		if err != nil {
			fmt.Printf("⚠️ Error processing %s: %v\n", imgPath, err)
			continue
		}
		if text != "" {
			allText = append(allText, fmt.Sprintf("--- %s ---\n%s", imgPath, text))
		} else {
			fmt.Printf("⚠️ No text extracted from %s\n", imgPath)
		}
	}

	if len(allText) == 0 {
		fmt.Println("❌ No text extracted from any image.")
		os.Exit(1)
	}

	combinedText := strings.Join(allText, "\n")
	fmt.Println("\n📖 Extracted text:")
	if len(combinedText) > 500 {
		fmt.Println(combinedText[:500] + "...")
	} else {
		fmt.Println(combinedText)
	}

	var err error
	if *format == "docx" {
		err = exportDocx(combinedText, *output)
		fmt.Println("\n📄 Exporting to DOCX... ✅")
	} else {
		err = exportTxt(combinedText, *output)
		fmt.Println("\n📄 Exporting to TXT... ✅")
	}
	if err != nil {
		fmt.Printf("Error exporting: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("File saved: %s\n", *output)
}
